rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+
+Status: Draft
+
+Authors:
+
+* [David Moravek](https://github.com/dmvk)
+
+References:
+
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity
+* https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go
+
+## Motivation
+
+To utilize SlateDB as a fully-featured state backend for a stream processor, it must support three key types of state:
+
+1. **Value**: Latest value or aggregating state.
+2. **List / Bag / Buffering**: State that stores multiple items for later processing.
+3. **Map**: Key-value state for fine-grained data.
+
+State-of-the-art implementations built on RocksDB often leverage the [Merge Operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator) to efficiently manage types 1 and 2. A prime example is list state, where elements are appended continuously until triggered by an event like a window firing. While this method may not always be the absolute optimal solution, it serves as a solid foundation for future iterations and enhancements.
+
+## Goals
+The primary goal is to introduce a new user-facing record type (**Merge Operand**) and a **Merge Operator** interface to SlateDB, allowing applications to bypass the traditional read/modify/update cycle in performance-critical situations where computation can be expressed using an [associative](https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity) operator.
+
+**Supporting non-associative operators is not a goal of this RFC**. The assumption is that associative operators can cover the majority of use cases while keeping the user-facing API simple and significantly reducing the scope. [Badger](https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go) has adopted the same limitation.
+
+## Public API
+
+We will introduce two new methods for writing Merge Operands into the database. They follow the same structure as existing put methods.
+
+```rust
+impl Db {
+    
+    pub async fn merge(
+        &self,
+        key: &[u8],
+        value: &[u8]
+    ) -> Result<(), SlateDBError> {
+	    ...
+    }
+
+    pub async fn merge_with_options(
+        &self,
+        key: &[u8],
+        value: &[u8],
+        merge_opts: &MergeOptions,
+        write_opts: &WriteOptions,
+    ) -> Result<(), SlateDBError> {
+        ...
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct MergeOptions {
+    pub ttl: Ttl,
+}
+```
+We also need to introduce a new trait that will allow users to implement custom merging logic.
+
+```rust
+trait MergeOperator {
+
+    pub fn name(&self) -> String;
+
+    pub fn merge(
+        &self,
+        key: Bytes,
+        existing_value: Bytes,
+        value: Bytes
+    ) -> Result<Bytes, MergeOperatorError>;
+}
+```
+
+Initially, there will be a single optional instance of the merge operator per database. The user needs to ensure that the compactor and writer run with the same merge operator to get predictable results.
```

Comment:
This should default to None because we can't assume desired merging logic.

I think there are two cases where we could encounter this:

**1. During initial write**

During initial write we have an option to simply refuse the record and push the error handling on user, but it would be rather tricky to implement (it's sort of the same case as if MO returns error) for two reasons:
- WriteBatch API: It should be atomic and we could have already persisted some parts of it.
- We can only perform best effort validation
  - We'd need to check against memtable before persisting to WAL (~ can we merge with records in memtable?). This would be prone to race conditions.
  - We shouldn't check against SST files for performance reasons.
  
This tells me that we should accept writes even if we can't perform merge right now and just postpone error handling until read operation 🤔 The tricky part is that we'd need to support multiple records for the same key in memtable.
  
**2. After record has been written (compactions, read)

Here we simply must leave the record intact instead of dropping it.

Reply:
I hadn't even given error handling much thought beyond the initial point, but this is a good line of thought. I'm in agreement that it makes sense to allow storing multiple merge values in the same SST for this type of situation.

This makes things... interesting. Some things we should flesh out in the RFC that come to mind:
- Ordering -- we probably want to make sure that within an SST they're ordered by something deterministic (`seqnum` makes sense). MOs should be associative, but I think forcing them to also be commutative is an extra burden. So long as we maintain ordering then we're good there. 
- Tombstones - I think this is independent of multiple values per key, but we should make sure that a delete + a write keeps the tombstone and doesn't "compact" it away. with normal writes a tombstone doesn't need to be kept if a more recent value shows up. There might be some other weird interleaving that make things complicated.
- Bloom filters. I assume we should mark a key present in an SST if there's any merge value in it, so that's pretty straightforward, but good to call out

Another way to do it is to have a native concept of an array of values for a merge value, so that we can store more than one merge value for a single key. This could make it a little easier to reason about. Basically we change the API for `KeyValueDeletable` to be:
```rust
KeyValueDeletable {
  Value(Bytes),
  Merged(&[Bytes]),
  Tombstone,
}
```

Reply:
Also related: https://github.com/slatedb/slatedb/issues/274

Reply:
> `Merged(&[Bytes]),`

I like this 👍

> Ordering

Ordering by seq id makes total sense.

> delete + a write keeps the tombstone

Here is more complete example of how I think about this.

```
tombstone + value => value
tombstone + merge => value
merge + merge => merge
```

> Bloom filters. I assume we should mark a key present in an SST if there's any merge value in it, so that's pretty straightforward, but good to call out

Yes. We need to do that, otherwise we might get incorrect results. The information bloom filter gives you is basically "the key might be present in a dataset" or "the key is 100% not present in the dataset". We need to avoid the 2nd case. I'll call it out in the RFC.

Reply:
> WriteBatch API: It should be atomic and we could have already persisted some parts of it.
We can only perform best effort validation
We'd need to check against memtable before persisting to WAL (~ can we merge with records in memtable?). This would be prone to race conditions.
We shouldn't check against SST files for performance reasons.
This tells me that we should accept writes even if we can't perform merge right now and just postpone error handling until read operation 🤔 The tricky part is that we'd need to support multiple records for the same key in memtable.

I don't think the atomicity is a concern necessarily. Atomicity is guaranteed by forcing all the writes into a single write batch into one WAL or L0 SST (depending on if wal is enabled or not). So you could do the validation of the batch before writing it, and then write it.

That being said, I think forcing a read on each merge write is prohibitively expensive as you say (especially if we need to include that in the atomic write) and rules out that approach. So I'm on board with either accepting merge operations even if the merge is None and deferring the error or just rejecting any merge when the operator is None. For the former approach, I think for snapshot isolation we'll need to include multiple versions of a key in a memtable anyway (e.g. a range scan with a snapshot tied to seq num N shouldn't observe a write with seq num > N), so there's already a good reason to support multiple versions of a key in the memtable.

Reply:
>  For the former approach, I think for snapshot isolation we'll need to include multiple versions of a key in a memtable anyway (e.g. a range scan with a snapshot tied to seq num N shouldn't observe a write with seq num > N)

Makes sense. I think this brings us back to `Merged(Bytes)` from the `Merged(&[Bytes])` proposed above. 🤔

I think most layers except for memtable should already support multiple values. Is there any in-progress work already for closing the gap?

Reply:
I've projected this into RFC, hope I didn't miss anything 🙈